[FSDP2] reset FSDPParam.sharded_param in lazy_init

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132954
* #133059

motivated by FSDP2 + DoRA https://github.com/pytorch/pytorch/issues/132721

after meta init, we need a user-defined function to move DoRALinear.magnitude from device=meta to device=cuda
The problem is how to trigger reset_sharded_param or _apply to update FSDPParam. Otherwise lazy_init complains that DoRALinear.magnitude are still on device=meta

credit to @awgu for chasing after DDP lazy_init to unblock the PR


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

weifengpy committed:
[FSDP2] reset FSDPParam.sharded_param in lazy_init

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -191,6 +191,12 @@ def _init_mp_dtypes(self) -> None:
 
     def lazy_init(self):
         # Lazy init should be idempotent
+        # User may initialize params again after meta init.
+        # For example, we initialize DoRA magnitudes base on
+        # other parameters loaded from the state dict.
```
(IIUC, it is not specific to meta init or not.)

awgu reviewed (approved):
Thanks!

weifengpy commented:
the CI error is relevant and I want to discuss my minimized repro with @awgu https://github.com/pytorch/pytorch/issues/132721

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -191,6 +191,12 @@ def _init_mp_dtypes(self) -> None:
+        # User may initialize params again after meta init.
+        # For example, we initialize DoRA magnitudes base on
+        # other parameters loaded from the state dict.
+        with torch.no_grad():
+            for fsdp_param in self.fsdp_params:
+                fsdp_param.reset_sharded_param()
```
I think to fix https://github.com/pytorch/pytorch/issues/132978
```suggestion
        if self.is_sharded:
            with torch.no_grad():
                for fsdp_param in self.fsdp_params:
                    fsdp_param.reset_sharded_param()
```

weifengpy replied:
just tested and the gap is still in DDP(FSDP).  FSDPModule params are "unsharded" to `torch.Tensor` already before calling `.unshard`, but `FSDPParamGroup.is_sharded` is still `True` 

https://github.com/pytorch/pytorch/blob/a9954d22f89677fc22fb9764f1edb0818538478f/test/distributed/_composable/fsdp/test_fully_shard_comm.py#L1066

weifengpy committed:
Update on "[FSDP2] reset FSDPParam.sharded_param in lazy_init"


motivated by FSDP2 + DoRA https://github.com/pytorch/pytorch/issues/132721

after meta init, we need a user-defined function to move DoRALinear.magnitude from device=meta to device=cuda
The problem is how to trigger reset_sharded_param or _apply to update FSDPParam. Otherwise lazy_init complains that DoRALinear.magnitude are still on device=meta


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] reset FSDPParam.sharded_param in lazy_init"


motivated by FSDP2 + DoRA https://github.com/pytorch/pytorch/issues/132721

after meta init, we need a user-defined function to move DoRALinear.magnitude from device=meta to device=cuda
The problem is how to trigger reset_sharded_param or _apply to update FSDPParam. Otherwise lazy_init complains that DoRALinear.magnitude are still on device=meta


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] reset FSDPParam.sharded_param in lazy_init"


motivated by FSDP2 + DoRA https://github.com/pytorch/pytorch/issues/132721

after meta init, we need a user-defined function to move DoRALinear.magnitude from device=meta to device=cuda
The problem is how to trigger reset_sharded_param or _apply to update FSDPParam. Otherwise lazy_init complains that DoRALinear.magnitude are still on device=meta

credit to awgu for chasing after DDP lazy_init to unblock the PR


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -536,6 +536,41 @@ def test_fully_shard_multi_module_root(self):
+                    weight_norm = torch.linalg.norm(
+                        self.layer1.weight, dim=1
+                    ) + torch.linalg.norm(self.layer2.weight, dim=1)
+                model.weight_norm = nn.Parameter(weight_norm)
```
@ebsmothers  examples of DoRA.magnitude in case you want to remove hacky code
